clang/lib/AST/PreorderAST.cpp

```diff
@@ -0,0 +1,200 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements methods to create and manipulate an n-ary preorder
+//  abstract syntax tree which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/AST/PreorderAST.h"
+
+using namespace clang;
+
+void PreorderAST::Create(Expr *E, Node *N, Node *Parent) {
+  if (!E)
+    return;
+
+  if (!N)
+    N = new Node(Parent);
+
+  // If the root is null, the current node is the root.
+  if (!Root)
+    Root = N;
+
+  // If the parent is non-null, make sure that the current node is marked as a
+  // child of the parent. As a convention, we create left children first.
+  if (Parent) {
+    if (!Parent->Left)
+      Parent->Left = N;
+    else
+      Parent->Right = N;
+  }
+
+  E = Lex.IgnoreValuePreservingOperations(Ctx, E);
+
+  // If E is a variable, store it in the variable list for the current node.
+  if (DeclRefExpr *D = GetDeclOperand(E)) {
+    if (const auto *V = dyn_cast_or_null<VarDecl>(D->getDecl())) {
+      N->Vars.push_back(V);
+      return;
+    }
+  }
+
+  // If E is a constant, store it in the constant field of the current node and
+  // set the HasConst field.
+  llvm::APSInt IntVal;
+  if (E->isIntegerConstantExpr(IntVal, Ctx)) {
+    N->Const = IntVal;
+    N->HasConst = true;
+    return;
+  }
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
+    // Set the opcode for the current node.
+    N->Opc = BO->getOpcode();
+
+    Expr *LHS = BO->getLHS()->IgnoreParens();
+    Expr *RHS = BO->getRHS()->IgnoreParens();
+  
+    if (isa<BinaryOperator>(LHS))
+      // Create the LHS as the left child of the current node.
+      Create(LHS, N->Left, N);
+    else
+      // Create the LHS in the current node.
+      Create(LHS, N);
+  
+    if (isa<BinaryOperator>(RHS))
+      // Create the RHS as the right child of the current node.
+      Create(RHS, N->Right, N);
+    else
+      // Create the RHS in the current node.
+      Create(RHS, N);
+  
+    return;
+  }
+
+  // Currently, we only handle expression which are either variables or
+  // constants.
+  // TODO: Handle expressions that are non-variables and non-constants.
+  // Possibly, add a field to the node to represent such expressions.
+  SetError();
+}
+
+void PreorderAST::Sort(Node *N) {
+  if (Error)
+    return;
+
+  if (!N || !N->Vars.size())
+    return;
+
+  if (!N->IsOpCommutativeAndAssociative()) {
+    SetError();
+    return;
+  }
+
+  // Sort the variables in the node lexicographically.
+  llvm::sort(N->Vars.begin(), N->Vars.end(),
+             [](const VarDecl *V1, const VarDecl *V2) {
+               return V1->getQualifiedNameAsString().compare(
```

Comment:
Two VarDecls may have the same name, but represent different variables from different scopes.  In this case, your sort may be unstable and put them into a different orders for different trees, which would cause the comparison to fail.